[MRG] LogReg: Check if l1_ratio is None when using elasticnet

joseortiz3 opened the pull request:
LogisticRegression with `penalty='elasticnet'` requires the user specify the `l1_ratio` between the L1 and L2 penalties, but the code currently does not check for this. The default value for `l1_ratio` is `None`.

Currently, calling `fit()` with the default value for `l1_ratio` will result in an unhelpful error stating `unsupported operand type(s) for -: 'int' and 'NoneType'`. This unhelpful error message is raised in `fit()` on the line:

```python
fold_coefs_ = Parallel(n_jobs=self.n_jobs, verbose=self.verbose, prefer=prefer)(
```

With the proposed change, it will now be clear to the user that the error was not specifying `l1_ratio`.

#### Reference Issues/PRs
N/A

joseortiz3 committed:
[MRG] LogisticRegression: Check if l1_ratio is None when using ElasticNet

LogisticRegression with `penalty=elasticnet` requires the user specify the `l1_ratio` between the L1 and L2 penalties, but the code currently does not check for this. 

Currently, an unhelpful error stating "unsupported operand type(s) for -: 'int' and 'NoneType'" is raised when `fold_coefs_` is calculated later in `fit()`.

With the proposed change, it will be clear to the user that the error was not specifying `l1_ratio`.

jeremiedbb reviewed (commented):
Thanks for the PR @joseortiz3. Please add a non-regression test in ``test_logistic.py``

joseortiz3 committed:
Added test to check that error message is helpful

This test checks that the error that occurs when l1_ratio=None and penalty='elasticnet' is helpful, meaning it contains the word 'l1_ratio'.

The error message for LogisticRegressionCV was already helpful:

`ValueError: l1_ratios must be a list of numbers between 0 and 1; got (l1_ratios=None)`

With the previous commit, the error message for LogisticRegression will be, too.

joseortiz3 committed:
Added a blank line to match the code formatting

Added a blank line to match the formatting of the document, two blank lines between every test.

joseortiz3 committed:
Try to fix linter fail that says file would be reformatted

jeremiedbb commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -227,6 +227,15 @@ def test_check_solver_option(LR):
 
+@pytest.mark.parametrize("LR", [LogisticRegression, LogisticRegressionCV])
+def test_elasticnet_l1_ratio_err_helpful(LR):
+    # Check that the error message contains the word 'l1_ratio'.
+    with pytest.raises(Exception, match=r".*l1_ratio.*"):
+        # Perform a simple LogisticRegression to trigger the error.
+        model = LR(penalty="elasticnet", l1_ratio=None, solver="saga")
+        model.fit(np.array([[1, 2], [3, 4]]), np.array([0, 1]))
```
The model instanciation should be outside of the context to make sure the error is raised at fit
```suggestion
    # Check that an informative error message is raised when penalty="elasticnet"
    # but l1_ratio is not specified.
    model = LR(penalty="elasticnet")
    with pytest.raises(Exception, match="l1_ratio must be specified"):
        model.fit(np.array([[1, 2], [3, 4]]), np.array([0, 1]))
```

jeremiedbb reviewed (approved):
LGTM. Thanks @joseortiz3

jeremiedbb committed:
Update sklearn/linear_model/tests/test_logistic.py

jeremiedbb committed:
cln
